algoliasearch/client.go

```diff
@@ -48,195 +46,171 @@ func (c *Client) SetTimeout(connectTimeout int, readTimeout int) {
 
 // ListIndexes returns the list of all indexes belonging to this Algolia
 // application.
-func (c *Client) ListIndexes() (interface{}, error) {
-	return c.transport.request("GET", "/1/indexes", nil, read)
+func (c *Client) ListIndexes() (ListIndexesRes, error) {
+	res, err := c.transport.request("GET", "/1/indexes", nil, read)
+	if err != nil {
+		return ListIndexesRes{}, err
+	}
+
+	indexes := ListIndexesRes{}
+	err = json.Unmarshal(res, &indexes)
+
+	return indexes, err
 }
 
 // InitIndex returns an Index object targeting `indexName`.
-func (c *Client) InitIndex(indexName string) *Index {
-	return NewIndex(indexName, c)
+func (c *Client) InitIndex(indexName string) Index {
+	return *NewIndex(indexName, c)
 }
 
 // ListKeys returns all the API keys available for this Algolia application.
-func (c *Client) ListKeys() (interface{}, error) {
-	return c.transport.request("GET", "/1/keys", nil, read)
+func (c *Client) ListKeys() (ListKeysRes, error) {
+	keys := ListKeysRes{}
+	err := c.request(&keys, "GET", "/1/keys", nil, read)
+	return keys, err
 }
 
 // MoveIndex renames the index named `source` as `destination`.
 func (c *Client) MoveIndex(source string, destination string) (interface{}, error) {
-	return c.InitIndex(source).Move(destination)
+	index := c.InitIndex(source)
+	return index.Move(destination)
 }
 
 // CopyIndex duplicates the index named `source` as `destination`.
 func (c *Client) CopyIndex(source string, destination string) (interface{}, error) {
-	return c.InitIndex(source).Copy(destination)
+	index := c.InitIndex(source)
+	return index.Copy(destination)
 }
 
-// AddKey creates a new API key named `key` using the `acl` operation
-// restrictions, the `indexes` the key is limited to, the `validity` during
-// which the key is to be valid in seconds, `maxQueriesPerIPPerHour` and
-// `maxHitsPerQuery` as rate limiters.
-func (c *Client) AddKey(acl, indexes []string, validity int, maxQueriesPerIPPerHour int, maxHitsPerQuery int) (interface{}, error) {
-	body := map[string]interface{}{
-		"acl":                    acl,
-		"maxHitsPerQuery":        maxHitsPerQuery,
-		"maxQueriesPerIPPerHour": maxQueriesPerIPPerHour,
-		"validity":               validity,
-		"indexes":                indexes,
+// AddKey creates a new API key from the supplied `ACL` and the specified
+// optional parameters.
+func (c *Client) AddKey(ACL []string, params map[string]interface{}) (AddKeyRes, error) {
```

Comment:
I don't understand this signature. Couldn't you just require from the user that the ACL are already in the `params` map? So that you won't need to add it by yourself in your function afterwards.